### ISSUE TITLE
alternator: support for controlling service level

### ISSUE THREAD

**Issue by 851110504** [scylladb/scylladb#17089]
In the process of using Alternator authentication, I found that the service level has a significant impact on performance. However, I did not use this feature. Here is my testing process.
I only created one user. The table only contains the user I created and a default user.
![image](https://github.com/scylladb/scylladb/assets/100502050/892c2595-cc91-41ab-94cc-ffafc2999fe5)
Then I started a Scylla node for testing, using go-ycsb for read and write operations. Under low load conditions, I ran go-ycsb with just 1 thread. The data under these conditions, without any modifications, is as follows.
![image](https://github.com/scylladb/scylladb/assets/100502050/38ca6d5d-3dca-4367-87db-9b113a34a2df)
Then I saw a function in the source code that updates parameters based on the service level. When I commented out this function, I ran go-ycsb under the same conditions and the data is as follows.
![image](https://github.com/scylladb/scylladb/assets/100502050/4dc9a92b-51c3-4388-b05e-ec87f2d83b25)
I also tested the differences under high load conditions, running go-ycsb with 32 threads. The data for these two situations is as follows.
![image](https://github.com/scylladb/scylladb/assets/100502050/98b58949-6490-4e41-a7c8-12b161496efa)
![image](https://github.com/scylladb/scylladb/assets/100502050/9354b5a6-a734-4aec-8611-27bc25476aef)
It is clear that the ops have dropped significantly, which has also led to an increase in latency. However, I did not use the service level feature. In order to improve performance, I added a parameter to control whether to use the service level.

**Comment by nyh** [comment]
Just to clarify, you mentioned the words "high load", "low load", but did one of these tests reach 100% CPU usage on the server?

**Comment by 851110504** [comment]
> Just to clarify, you mentioned the words "high load", "low load", but did one of these tests reach 100% CPU usage on the server?

It did not reach 100% CPU utilization. Under low load, the CPU utilization was around 5%. Under high load, it was around 60%.

**Comment by nyh** [comment]
Hi @eliransin maybe you can please help me understand how this "service level" thing works. In 2379a25adea64f243f0144d9c75805c400b2f70b Piotr Sarna added putting "sl_controller" into the client_state. This user reports that this noticeably reduced performance (at least **latency**) - but why? What do the queries (reads and writes) which receive this "sl_controller" actually do differently than they would without this sl_controller? Could they make requests _slower_? Could they change some defaults and increase latencies?

@851110504 in your pull request https://github.com/scylladb/scylladb/pull/17055 you solved this performance degradation by making the passing of sl_controller optional, but  I would prefer to try to understand if we can make the default behavior and performance _with_ this sl_controller to match the one without it - if we could do that we wouldn't need this option. I'm not saying we can do that but before saying we can't I want to understand this code better.

**Comment by nyh** [comment]
> > Just to clarify, you mentioned the words "high load", "low load", but did one of these tests reach 100% CPU usage on the server?
> 
> It did not reach 100% CPU utilization. Under low load, the CPU utilization was around 5%. Under high load, it was around 60%.

My problem is that I can't tell if the "service level" problem just increased latency (perhaps because the default settings are wrong?) - or also reduced throughput. Those aren't the same thing. To show that throughput is reduced you would either need to reach 100% CPU usage in both regular and commented-out runs and show different ops, or, show that at a fixed ops-per-second, you use more CPU to achieve this fixed ops-per-second.

I have another question: You said "Then I saw a function in the source code that updates parameters based on the service level." - which function is that?

**Comment by 851110504** [comment]
> > > Just to clarify, you mentioned the words "high load", "low load", but did one of these tests reach 100% CPU usage on the server?
> > 
> > 
> > It did not reach 100% CPU utilization. Under low load, the CPU utilization was around 5%. Under high load, it was around 60%.
> 
> My problem is that I can't tell if the "service level" problem just increased latency (perhaps because the default settings are wrong?) - or also reduced throughput. Those aren't the same thing. To show that throughput is reduced you would either need to reach 100% CPU usage in both regular and commented-out runs and show different ops, or, show that at a fixed ops-per-second, you use more CPU to achieve this fixed ops-per-second.
> 
> I have another question: You said "Then I saw a function in the source code that updates parameters based on the service level." - which function is that?

![code](https://github.com/scylladb/scylladb/assets/100502050/bba1d509-618a-4fe8-b3a6-127a36fa2006)
This function is over here. Also, I have another question. The authentication of Alternator requires the password to be stored in plaintext in the database. I noticed that cqlsh uses encryption logic. Haven’t you optimized the authentication part of Alternator?

**Comment by nyh** [comment]
About the (in)security of storing the correct passwords on the Scylla nodes, and how it can be fixed, see https://github.com/scylladb/scylladb/issues/5206.

**Comment by 851110504** [comment]
> Hi @eliransin maybe you can please help me understand how this "service level" thing works. In [2379a25](https://github.com/scylladb/scylladb/commit/2379a25adea64f243f0144d9c75805c400b2f70b) Piotr Sarna added putting "sl_controller" into the client_state. This user reports that this noticeably reduced performance (at least **latency**) - but why? What do the queries (reads and writes) which receive this "sl_controller" actually do differently than they would without this sl_controller? Could they make requests _slower_? Could they change some defaults and increase latencies?
> 
> @851110504 in your pull request #17055 you solved this performance degradation by making the passing of sl_controller optional, but I would prefer to try to understand if we can make the default behavior and performance _with_ this sl_controller to match the one without it - if we could do that we wouldn't need this option. I'm not saying we can do that but before saying we can't I want to understand this code better.

![service_level](https://github.com/scylladb/scylladb/assets/100502050/305ccb33-9b3a-4a7e-b5f3-633a48a2a8b0)
The main impact should be on the query_granted and find_service_level methods. The query_granted method recursively queries which roles the current user has, and then the find_service_level method iterates through the retrieved roles to aggregate service levels. Since I haven't set any service levels, in my execution flow, I return directly after getting slo_opt. Additionally, because I only have one user stored in the database, I don't know why there is still a significant impact.

**Comment by nyh** [comment]
> Additionally, because I only have one user stored in the database, I don't know why there is still a significant impact.

That's exactly what I don't understand. I can understand that all this "service level" checking adds a few microseconds of ifs that aren't important, but I don't understand why it has a **significant** impact on performance. 

If I understand correctly, role_manager.query_granted() does need to interact with an on-disk table which can be slow, but it should be cached so actually reading from disk should be exceedingly rare, if any. Maybe it's not rare enough? Or maybe your test is too sensitive to the latency increase caused from the first requests, or something?

I'd really like to understand what's going on before adding a configuration option to bypass a piece of code we just didn't understand.

@ptrsmrn who do we have who's familiar with service levels and can look at this?

**Comment by 851110504** [comment]
> > Additionally, because I only have one user stored in the database, I don't know why there is still a significant impact.
> 
> That's exactly what I don't understand. I can understand that all this "service level" checking adds a few microseconds of ifs that aren't important, but I don't understand why it has a **significant** impact on performance.
> 
> If I understand correctly, role_manager.query_granted() does need to interact with an on-disk table which can be slow, but it should be cached so actually reading from disk should be exceedingly rare, if any. Maybe it's not rare enough? Or maybe your test is too sensitive to the latency increase caused from the first requests, or something?
> 
> I'd really like to understand what's going on before adding a configuration option to bypass a piece of code we just didn't understand.
> 
> @ptrsmrn who do we have who's familiar with service levels and can look at this?

![service_level](https://github.com/scylladb/scylladb/assets/100502050/ea40ac4a-6c98-4da7-9001-059f58dacd37)
In role_manager.query_granted(), there should be a cache in place. However, the latency increase doesn't seem to be caused only by the first request; it appears to persist.

**Comment by ptrsmrn** [comment]
> @ptrsmrn who do we have who's familiar with service levels and can look at this?

@nyh I think @Jadw1 would be the person.

**Comment by kostja** [comment]
@851110504 could you please provide the exact steps to reproduce the problem?

1) Which branch/version of go-ycsb are you using?
2) Any specific build steps?
3) Scylla.yaml configuration file in use?
4) command line parameters to run the benchmark and get the numbers
5) Anything else you deep necessary to make sure we can follow your exact steps to reproduce the problem.

Thanks!

**Comment by kostja** [comment]
@851110504 any specific start-up options for go-ycsb?

**Comment by 851110504** [comment]
> @851110504 any specific start-up options for go-ycsb?

**dynamodb.properties**
dynamodb.tablename = test_ycsb
dynamodb.primarykey = bi
dynamodb.hashkey.name = obj
dynamodb.hashkey.value = single_partion_mutipart_test_001
dynamodb.rc.units = 10
dynamodb.wc.units = 10
dynamodb.ensure.clean.table = false
dynamodb.endpoint = http://172.17.0.7:11000
dynamodb.region = us-east-1
dynamodb.consistent.reads = false
dynamodb.delete.after.run.stage = false
dynamodb.request.timeout.ms = 10000
dynamodb.maxretry = 3
dynamodb.getitem.errorlog = getitem_error.log
dynamodb.putitem.errorlog = putitem_error.log
dynamodb.table.deletion.protection.enabled = false

verifyrow.logfile = verifyrow.log

dynamodb.disableValidateResponseChecksum = true
dynamodb.hashKeyNum = 1000000

**workloada**
recordcount=1000000000
operationcount=1000000000
workload=core
fieldcount=1
fieldlength=1000

readallfields=true
writeallfields=true
dataintegrity=false

readproportion=0.5
deleteproportion=0
updateproportion=0
scanproportion=0
insertproportion=0.5
readmodifywriteproportion=0

requestdistribution=uniform
core_workload_insertion_retry_limit=0

**Startup Command**
./go-ycsb load dynamodb -P dynamodb.properties -P workloads/workloada --threads 32
./go-ycsb run dynamodb -P dynamodb.properties -P workloads/workloada --threads 32

**Comment by dani-tweig** [comment]
@nyh - area/service levels label was added.
Thank you.

**Comment by nyh** [comment]
> @nyh - area/service levels label was added. Thank you.

Thanks. As you may have noticed I did a short effort to put this tag on bugs on a few issues that could use it. I'm sure I didn't find everything.

**Comment by kostja** [comment]
The current master has significantly reworked authentication and service levels. You may want to re-test with our nightly build available at https://downloads.scylladb.com/unstable/scylla/master/

**Comment by nyh** [comment]
> The current master has significantly reworked authentication and service levels. You may want to re-test with our nightly build available at https://downloads.scylladb.com/unstable/scylla/master/

If this issue still exists (and I assume it did exist - I don't think @851110504 imagined it), it may be Alternator performance dropped on the floor, and I think it's important that someone works on it.

Note that I always figured that authentication costs in performance - after all we do all sort of signature calculations that we wouldn't do without authentication. But what @851110504 suggests is that authentication wasn't the big problem (and improvements to the authentication code wouldn't have helped) - the big problem had to do with "service levels" which are completely irrelevant (because he never defined any service levels).

**Comment by 851110504** [comment]
> > The current master has significantly reworked authentication and service levels. You may want to re-test with our nightly build available at https://downloads.scylladb.com/unstable/scylla/master/
> 
> If this issue still exists (and I assume it did exist - I don't think @851110504 imagined it), it may be Alternator performance dropped on the floor, and I think it's important that someone works on it.
> 
> Note that I always figured that authentication costs in performance - after all we do all sort of signature calculations that we wouldn't do without authentication. But what @851110504 suggests is that authentication wasn't the big problem (and improvements to the authentication code wouldn't have helped) - the big problem had to do with "service levels" which are completely irrelevant (because he never defined any service levels).

Yes, authentication also has an impact on performance, mainly in the generation of v4 signatures. However, the bigger impact comes from service levels. Even if service levels are not defined, the performance degradation is still significant.

**Comment by nuivall** [comment]
I suspect this is because `maybe_update_per_service_level_params` is called on each connection start with cql, which unfortunately for alternator means every request. And this function is way too heavy.
- It doesn't use specialized permissions cache which we normally use on cql hot path (because it's not checking permissions it's doing something else...)
- It does use generic row cache
- Maybe the code is heavy by itself (we generally don't profile non hot paths so we wouldn't notice [as code author clearly didn't consider it to be on a hot path], worth to measure instructions count or allocations for this)
- And most importantly perhaps due to nature of this workload (overload or near overload?) the cache hit ratio is very low and the problem intensifies

The change on master may help depending on your setup (especially if you have mutli node with improper auth RF setting) but I don't know if this is the case here.


BTW the `cql3::query_processor::cache_internal::yes` you saw is not exactly about caching results, it means it will use (cached) prepared statement.

**Comment by nyh** [comment]
I'll try to work on this issue now. My preference is _not_ to add an option to **disable** service levels, but rather make it a no-op (as it should be) when not used and certainly not have any large performance impact.

**Comment by avikivity** [comment]
A lighter-weight low-level switching function is coroutine::switch_to(). It avoids extra allocations.

**Comment by nyh** [comment]
Even without sophisticated benchmarking or profiling, Just adding a timing of `maybe_update_per_service_level_params` in alternator/server.cc 
```c++
    start_time = std::chrono::steady_clock::now();
    co_await client_state.maybe_update_per_service_level_params();
    slogger.warn("NYH maybe_update_per_service_level_params time {}", std::chrono::steady_clock::now() - start_time);
```
and running some test (to get requests going) reproduces the problem that @851110504 complained about:

It shows that this hoped no-op function takes between 50 and 100 microseconds for each call. To put things in persepctive, the verify_signature function that I was worried about, takes about 25 microseconds. So definitely we have a problem.

**Comment by nuivall** [comment]
> It shows that this hoped no-op function

that's only the case a) SL not used, for case b) SL used it's not nop but also needs to be fast

**Comment by nyh** [comment]
It turns out `client_state.maybe_update_per_service_level_params()` has two slow parts: `role_manager.query_granted()` and `_sl_controller->find_service_level()`. The first is slow because it uses `find_record()` (standard_role_manager.cc) which always performs a query in the roles table. The second is slow because it calls `get_attribute()` (also standard_role_manager.cc), which also runs a query.

I see two options how to solve this:

1. The entire standard_role_manager.cc implementation should be modified to **cache** results of recent searches, so we don't need to perform internal queries. These internal queries are cached using our normal row caches etc., but still, the entire query machinary is probably too slow. The problem with this is that this code has nothing to do with Alternator and I'm worried I'll be stepping on other people's work on service levels. CC @piotrd @nuivall @Jadw1 
2. In Alternator, we could cache "client_state" for frequently used users, without touching any of the implementation code that builds this client_state and without really caring how it works.

Because I'm afraid to touch the bulk of the service level code to use option 1, I'm tempted to use option 2, although it won't be pretty. I'll need to use a cache (loading_cache?) which keeps the result only for a few seconds because I wouldn't know when to refresh it, and only a small size because I wouldn't know how to size it properly.

That being said, I'm still puzzled about something, maybe some of the people listed above or @kostja or @mykaul can help me understand:. I thought we do cache permissions - we have permissions_validity_in_ms, permissions_update_interval_in_ms, permissions_cache_max_entries for configuring this cache. Am I to understand now that "permissions" does not include "roles"? Should standard_role_manager.cc use a similar approach to cache its work, just like permissions do? If we did that, Alternator will benefit from it immediately.

**Comment by mykaul** [comment]
I don't know what we cache, but I do recall the default cache validity is only 2 seconds or something, not very effective.

**Comment by nyh** [comment]
> I don't know what we cache, but I do recall the default cache validity is only 2 seconds or something, not very effective.

Well, if you have 1000 queries per second per shard, and you only do those silly lookups only once every 2 seconds instead of on every query, this is a 2000-fold reduction of the overhead of these lookups.

By the way, I see the default permissions_validity_in_ms is 10 seconds, not 2 seconds - but there is a 2-second default for permissions_update_interval_in_ms. In any case, as I noted above it seems that caching only applies to "permissions", not to "roles" :-(

**Comment by nuivall** [comment]
> That being said, I'm still puzzled about something, maybe some of the people listed above or @kostja or @mykaul can help me understand:. I thought we do cache permissions - we have permissions_validity_in_ms, permissions_update_interval_in_ms, permissions_cache_max_entries for configuring this cache. Am I to understand now that "permissions" does not include "roles"? Should standard_role_manager.cc use a similar approach to cache its work, just like permissions do? If we did that, Alternator will benefit from it immediately.

Yes, it only caches `auth::get_permissions` nothing else. I talked about it during summit presentation, on the last slide as future improvement.

But one thing I don't understand nor I think your benchmark measurement answered, why row cache doesn't help here? Is it merely too many layers below so that overhead is already too high even without disc io? Or is it somehow not used?

How 100 microseconds translates to tps drop? I think first comment demonstrates 15% drop. Were you able to reproduce this number?

Also if it doesn't include io (or does it?) what takes 0.1ms ? This seems a lot, is it maybe badly written code somewhere? 

Those questions are important because one of potential cache v2 for auth implementation could be re-using row cache by changing eviction policy for auth tables (I think you even suggested that in some ticket or mail).

**Comment by nyh** [comment]
>     2. In Alternator, we could cache "client_state" for frequently used users, without touching any of the implementation code that builds this client_state and without really caring how it works.

It turns out that client_state is deliberately non-copyable, I can't even save a copy of a client_state built by a function I want to avoid changing (maybe_update_per_service_level_params()), and I will only be able to use a cache if the client state is read-only after creation and can be reused. Maybe because we don't yet insert the client endpoint address into the client_state, it can indeed reuse it. But it will be super ugly.

In general, I think I stepped on a nest of hornets :-( As Marcin noted above, maybe_update_per_service_level_params() is only used in two places: 1. In CQL it is used once per connection on which later many requests pass, and 2. in Alternator, for every single request. Whereas in CQL nobody cares about the performance of the connection initialization, in Alternator, the overhead is much more disturbing. So it seems this entire code was written without regard to this performance  - and now I need to decide what to do about it, but to be honest I'm completely unfamiliar with this code and a bit afraid to botch it up. Especially when we have other repositories modifying the same code to do other things, and I'm not familiar with that too.

**Comment by nyh** [comment]
> Yes, it only caches `auth::get_permissions` nothing else. I talked about it during summit presentation, on the last slide as future improvement.

Oh :-(
 
> But one thing I don't understand nor I think your benchmark measurement answered, why row cache doesn't help here? Is it merely too many layers below so that overhead is already too high even without disc io? Or is it somehow not used?

The row cache does help somewhat. For example, the very first find_service_level() (the second half of the slow lookuo) call takes 80 microseconds, and the second takes 20 microseconds. So the first call is even worse. But the second is not good enough. I guess there is a huge amount of overhead involved in a CQL queries.

When I say "huge" I am exaggerating, of course. These role queries are faster than the user's query, and don't double the cost of the request. But since we do 2 (or more?) of these internal queries, and each has a noticeable overhead, the resulting overhead is noticeable to say the least. And since "service levels" by default do absolutely nothing, it's a shame that they noticeably hurt performance. So @851110504 in his PR wanted to add an option to completely disable service levels. I would have preferred fixing the performance instead of adding an option to officially "disable" a feature that the user wasn't using anyway.

> How 100 microseconds translates to tps drop? I think first comment demonstrates 15% drop. Were you able to reproduce this number?

My "benchmark" involves (as I noted above) printing wall-clock time in the code and driving it with a random test/alternator test which does a few requests sequentially. It's not a real measure of performance and I can't give numbers like "15%". But I can see which function calls take time and which don't, and the main thing I wanted to understand was whether the silly function that I thought was a no-op really takes more time than the complicated signature-calculating function - and the answer is clearly "yes".

When you add to your benchmark code support for authentication, I can use it to do better benchmarking :-)

> Also if it doesn't include io (or does it?) what takes 0.1ms ? This seems a lot, is it maybe badly written code somewhere?

This 0.05ms (0.1ms was the maximum I ever saw) includes at least two separate CQL queries and a bunch of code around it, which I guess is pretty heavy compared to the nothing we would have expected it to do. Avi also suggested above there is some expense in switching scheduling group, although I didn't see this code.

> Those questions are important because one of potential cache v2 for auth implementation could be re-using row cache by changing eviction policy for auth tables (I think you even suggested that in some ticket or mail).

If I suggested this, then maybe I didn't realize how much a full-fledged CQL query is slower than a hash table lookup. I think we need to do some micro-benchmarks - just run the different types of lookups in a loop and see how fast or slow each one is. If (and I'm just making numbers up) a full-fledged CQL query from row cache takes 50 microseconds but a hash table lookup only takes 1 microsecond, we need to think hard if it's good enough.

**Comment by avikivity** [comment]
> Even without sophisticated benchmarking or profiling, Just adding a timing of `maybe_update_per_service_level_params` in alternator/server.cc
> 
> ```c++
> start_time = std::chrono::steady_clock::now();
> co_await client_state.maybe_update_per_service_level_params();
> slogger.warn("NYH maybe_update_per_service_level_params time {}", std::chrono::steady_clock::now() - start_time);
> ```
> 
> and running some test (to get requests going) reproduces the problem that @851110504 complained about:
> 
> It shows that this hoped no-op function takes between 50 and 100 microseconds for each call. To put things in persepctive, the verify_signature function that I was worried about, takes about 25 microseconds. So definitely we have a problem.

Are you measuring latency or throughput?

**Comment by nyh** [comment]
> Are you measuring latency or throughput?

My measurement was a trivial std::chrono timing around a function call, so it's **latency**. It was simply much easier than writing a real DynamoDB API loader which will load the server. However, in this case the latency/throughput distinctions houldn't matter: after the first call we expect the row cache to be filled with the searched data so there would be no sleeps, and the 50 microseconds or whatever is CPU usage. Of course I didn't prove it. But in any case, ideally, this function should have been something as quick as a hash-table lookup, not something that takes 50-100 microseconds.

**Comment by nyh** [comment]
> > That being said, I'm still puzzled about something, maybe some of the people listed above or @kostja or @mykaul can help me understand:. I thought we do cache permissions - we have permissions_validity_in_ms, permissions_update_interval_in_ms, permissions_cache_max_entries for configuring this cache. Am I to understand now that "permissions" does not include "roles"? Should standard_role_manager.cc use a similar approach to cache its work, just like permissions do? If we did that, Alternator will benefit from it immediately.
> 
> Yes, it only caches `auth::get_permissions` nothing else. I talked about it during summit presentation, on the last slide as future improvement.

Hi @nuivall, I want to make it clear here (or perhaps in a new issue) what is the plan here, and then we can decide **when** we want to execute this plan, or **who** will do it. Is the plan to add caching in `client_state.maybe_update_per_service_level_params()` or the lower-level `role_manager.query_granted()` and `_sl_controller->find_service_level()`? Is the plan to use the same sort of mechanisms that are used today for caching permissions - or the plan is to change both of them to use some new mechanism?

> Those questions are important because one of potential cache v2 for auth implementation could be re-using row cache by changing eviction policy for auth tables (I think you even suggested that in some ticket or mail).

If I understand correctly, my tests were small and had no reason to evict anything from the auth tables, so all the slowness I saw came from all the machinery around CQL queries, and not from the disk access. It's just that for every real request, we add two more internal CQL requests, so if the real request is fast (e.g., also a read from the cache) the slowdown is noticeable.

**Comment by nyh** [comment]
@nuivall @851110504 I opened a new issue https://github.com/scylladb/scylladb/issues/18738 about speeding up the service-level lookups by caching. I'm leaving this issue open because this issue specifically requested a _different_ feature - a configurable option to not do service-level options at all, but I believe that if https://github.com/scylladb/scylladb/issues/18738 is done these lookups will be so quick that nobody will ever want to bother to disable them, so this issue will become moot and could be closed.

**Comment by avikivity** [comment]
> > Are you measuring latency or throughput?
> 
> My measurement was a trivial std::chrono timing around a function call, so it's **latency**. It was simply much easier than writing a real DynamoDB API loader which will load the server. However, in this case the latency/throughput distinctions houldn't matter: after the first call we expect the row cache to be filled with the searched data so there would be no sleeps, and the 50 microseconds or whatever is CPU usage. Of course I didn't prove it. But in any case, ideally, this function should have been something as quick as a hash-table lookup, not something that takes 50-100 microseconds.

How do you know the hash lookup took 50 microseconds? Maybe some unrelated tasks ran in between.

**Comment by nuivall** [comment]
> > > That being said, I'm still puzzled about something, maybe some of the people listed above or @kostja or @mykaul can help me understand:. I thought we do cache permissions - we have permissions_validity_in_ms, permissions_update_interval_in_ms, permissions_cache_max_entries for configuring this cache. Am I to understand now that "permissions" does not include "roles"? Should standard_role_manager.cc use a similar approach to cache its work, just like permissions do? If we did that, Alternator will benefit from it immediately.
> > 
> > 
> > Yes, it only caches `auth::get_permissions` nothing else. I talked about it during summit presentation, on the last slide as future improvement.
> 
> Hi @nuivall, I want to make it clear here (or perhaps in a new issue) what is the plan here, and then we can decide **when** we want to execute this plan, or **who** will do it. Is the plan to add caching in `client_state.maybe_update_per_service_level_params()` or the lower-level `role_manager.query_granted()` and `_sl_controller->find_service_level()`? Is the plan to use the same sort of mechanisms that are used today for caching permissions - or the plan is to change both of them to use some new mechanism?
> 

Step 0 of my plan is to do profiling and try to understand what exactly is slow here. When: it is scheduled after I manage to merge https://github.com/scylladb/scylladb/pull/17910

**Comment by nyh** [comment]
> How do you know the hash lookup took 50 microseconds? Maybe some unrelated tasks ran in between.

It's possible, and I agree my methodology wasn't bulletproof. But I did run a single-threaded test with hundreds of different Alternator requests (basically, the Alternator test suite), and all of them had this delay - it wasn't some random event that some other work got in the middle.

**Comment by kostja** [comment]
@nyh @nuivall is not working on service levels.

**Comment by piodul** [comment]
After https://github.com/scylladb/scylladb/pull/19085 we no longer query from the system tables, the service levels are cached in an `std::unordered_map`. It should be much faster than querying system tables, even if their contents are in the cache.

@851110504 would you be able to retest with a nightly version? The fix will be available in 6.2 but it's not released yet.

**Comment by nyh** [comment]
@851110504 as I mentioned above, I think we should close this issue - the reason why you wanted to disable service levels in Alternator was the service level lookup was very slow. But now that this lookup is cached, it should be negligible and thus there is no reason why anyone might want to explicitly disable it.

If you agree, please close this issue. I would be even better if you can return to your benchmarks which showed bad performance once authentication is enabled - and check that the problem is indeed gone.
